File: credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
bool checkNumeric(string cn);
int checkCardType(string cn);
int checkLength(string cn);
int checkSum(string cn);

int main(void)
{
    string creditNum;
    do
    {
        creditNum = get_string("Enter Your Credit Card Number:");
    }
    while (checkNumeric(creditNum) == false);


    int ct = checkCardType(creditNum);

    if (ct == 2)
    {
        int v = checkSum(creditNum);
        if (v == 10)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (ct == 3)
    {
        int a = checkSum(creditNum);
        if (a == 10)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (ct == 4)
    {
        int mc = checkSum(creditNum);
        if (mc == 10)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (ct == 1)
    {
        printf("INVALID\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

bool checkNumeric(string cn)
{
    int l = strlen(cn);
    if (l == 0)
    {
        return false;
    }

    for (int i = 0; i < l; i++)
    {
        if (!isdigit(cn[i]))
        {
            return false;
        }
    }
    return true;
}

int checkCardType(string cn)
{
    int l = strlen(cn);
    if (cn[0] == '4') //Visa
    {
        if (l == 13 || l == 16)
        {
            return 2;
        }
    }
    else if (cn[0] == '3') //AmEx
    {
        if (cn[1] == '4' || cn[1] == '7')
        {
            if (l == 15)
            {
                return 3;
            }
        }
    }
    else if (cn[0] == '5') //MasterCard
    {
        if (cn[1] == '1' || cn[1] == '2' || cn[1] == '3' || cn[1] == '4' || cn[1] == '5')
        {
            if (l == 16)
            {
                return 4;
            }
        }
    }
    return 1;
}

int checkSum(string cn)
{
    int l = strlen(cn);
    long ccn = atol(cn);
    int xx = 0;
    int yy = xx;
    xx = ccn % 10;
    yy += xx;
    ccn /= 10;
    int x = 0;
    int z = 0;
    int y = x + z;

    for (int i = 0, j = (l / 2); i < j; i++)
    {
        z = 0;
        x = ccn % 10;
        x *= 2;
        if (x > 9)
        {
            z = x % 10;
            x /= 10;
        }
        y = x + y + z;
        ccn /= 10;
        xx = ccn % 10;
        yy += xx;
        ccn /= 10;
    }
    y += yy;

    if (y % 10 == 0)
    {
        return 10;
    }
    else
    {
        return 0;
    }
}

File: credit.py
from cs50 import get_string
import sys
import math


def checkNumeric(creditNum):

    if creditNum.isdigit():
        return True
    return False
    

def checkCardType(creditNum):
    l = len(creditNum)
    if creditNum[0] == '4':
        if l == 13 or l == 16:  # Visa
            return 2
        else:
            return 1
    elif creditNum[0] == '3':  # AmEx
        if creditNum[1] == '4' or creditNum[1] == '7':
            if l == 15:
                return 3
            else:
                return 1
    elif creditNum[0] == '5':  # MasterCard
        if creditNum[1] == '1' or creditNum[1] == '2' or creditNum[1] == '3' or creditNum[1] == '4' or creditNum[1] == '5':
            if l == 16:
                return 4
            else:
                return 1
    else:
        return 1


def checkSum(creditNum):
    l = len(creditNum)
    ccn = int(creditNum)
    xx = 0
    yy = xx
    xx = ccn % 10
    yy += xx
    ccn = math.floor(ccn / 10)
    x = 0
    z = 0
    y = x + z
    jj = l / 2
    math.floor(jj)
    j = int(jj)

    for i in range(j):
        z = 0
        x = ccn % 10
        x *= 2
        if x > 9:
            z = x % 10
            x = math.floor(x / 10)
        y = x + y + z
        ccn = math.floor(ccn / 10)
        xx = ccn % 10
        yy += xx
        ccn = math.floor(ccn / 10)
    y += yy

    if y % 10 == 0:
        return 10
    else:
        return 0


while True:
    creditNum = get_string("Enter Your Credit Card Number: ")
    if checkNumeric(creditNum):
        break
    print("Please enter a valid number")
ct = checkCardType(creditNum)
if ct == 1:
    print("INVALID")
if ct == 2:
    v = checkSum(creditNum)
    if v == 10:
        print("VISA")
    else:
        print("INVALID")
elif ct == 3:
    a = checkSum(creditNum)
    if a == 10:
        print("AMEX")
    else:
        print("INVALID")
elif ct == 4:
    mc = checkSum(creditNum)
    if mc == 10:
        print("MASTERCARD")
    else:
        print("INVALID")